File: packages/python-sdk/moru/volume_async/volume_api.py
"""Async VolumeApi class with static methods for volume operations."""

from typing import List, Optional, Tuple

from typing_extensions import Unpack

from moru.api import handle_api_exception
from moru.api.client_async import get_api_client
from moru.connection_config import ApiParams, ConnectionConfig
from moru.exceptions import NotFoundException
from moru.volume.types import FileInfo, VolumeInfo


class AsyncVolumeApi:
    """Base class for async volume API operations."""

    @staticmethod
    async def _create_volume(
        name: str,
        **opts: Unpack[ApiParams],
    ) -> VolumeInfo:
        """
        Create a new volume (idempotent).

        :param name: Volume name
        :return: Volume info
        """
        config = ConnectionConfig(**opts)
        api_client = get_api_client(config)
        client = api_client.get_async_httpx_client()

        response = await client.post(
            f"{config.api_url}/volumes",
            json={"name": name},
            timeout=config.request_timeout,
        )

        err = handle_api_exception(response)
        if err:
            raise err

        return VolumeInfo._from_api_response(response.json())

    @staticmethod
    async def _get_volume(
        volume_id_or_name: str,
        **opts: Unpack[ApiParams],
    ) -> VolumeInfo:
        """
        Get volume by ID or name.

        :param volume_id_or_name: Volume ID (vol_xxx) or name
        :return: Volume info
        """
        config = ConnectionConfig(**opts)
        api_client = get_api_client(config)
        client = api_client.get_async_httpx_client()

        response = await client.get(
            f"{config.api_url}/volumes/{volume_id_or_name}",
            timeout=config.request_timeout,
        )

        if response.status_code == 404:
            raise NotFoundException(f"Volume '{volume_id_or_name}' not found")

        err = handle_api_exception(response)
        if err:
            raise err

        return VolumeInfo._from_api_response(response.json())

    @staticmethod
    async def _list_volumes(
        limit: Optional[int] = None,
        next_token: Optional[str] = None,
        **opts: Unpack[ApiParams],
    ) -> Tuple[List[VolumeInfo], Optional[str]]:
        """
        List all volumes.

        :param limit: Maximum number of volumes to return
        :param next_token: Pagination token
        :return: Tuple of (volumes, next_token)
        """
        config = ConnectionConfig(**opts)
        api_client = get_api_client(config)
        client = api_client.get_async_httpx_client()

        params = {}
        if limit is not None:
            params["limit"] = limit
        if next_token is not None:
            params["nextToken"] = next_token

        response = await client.get(
            f"{config.api_url}/volumes",
            params=params,
            timeout=config.request_timeout,
        )

        err = handle_api_exception(response)
        if err:
            raise err

        volumes = [VolumeInfo._from_api_response(v) for v in response.json()]
        result_next_token = response.headers.get("x-next-token")

        return volumes, result_next_token

    @staticmethod
    async def _delete_volume(
        volume_id_or_name: str,
        **opts: Unpack[ApiParams],
    ) -> bool:
        """
        Delete a volume.

        :param volume_id_or_name: Volume ID (vol_xxx) or name
        :return: True if deleted, False if not found
        """
        config = ConnectionConfig(**opts)
        api_client = get_api_client(config)
        client = api_client.get_async_httpx_client()

        response = await client.delete(
            f"{config.api_url}/volumes/{volume_id_or_name}",
            timeout=config.request_timeout,
        )

        if response.status_code == 404:
            return False

        err = handle_api_exception(response)
        if err:
            raise err

        return True

    @staticmethod
    async def _list_files(
        volume_id: str,
        path: str = "/",
        limit: Optional[int] = None,
        next_token: Optional[str] = None,
        **opts: Unpack[ApiParams],
    ) -> Tuple[List[FileInfo], Optional[str]]:
        """
        List files in a volume.

        :param volume_id: Volume ID (vol_xxx)
        :param path: Directory path to list
        :param limit: Maximum number of files to return
        :param next_token: Pagination token
        :return: Tuple of (files, next_token)
        """
        config = ConnectionConfig(**opts)
        api_client = get_api_client(config)
        client = api_client.get_async_httpx_client()

        params = {"path": path}
        if limit is not None:
            params["limit"] = limit
        if next_token is not None:
            params["nextToken"] = next_token

        response = await client.get(
            f"{config.api_url}/volumes/{volume_id}/files",
            params=params,
            timeout=config.request_timeout,
        )

        if response.status_code == 404:
            raise NotFoundException(f"Volume '{volume_id}' not found")

        err = handle_api_exception(response)
        if err:
            raise err

        data = response.json()
        files = [FileInfo._from_api_response(f) for f in data.get("files", [])]
        result_next_token = data.get("nextToken")

        return files, result_next_token

    @staticmethod
    async def _upload_file(
        volume_id: str,
        path: str,
        content: bytes,
        **opts: Unpack[ApiParams],
    ) -> int:
        """
        Upload file content to volume.

        :param volume_id: Volume ID (vol_xxx)
        :param path: Destination path in volume
        :param content: File content as bytes
        :return: Size of uploaded file
        """
        config = ConnectionConfig(**opts)
        api_client = get_api_client(config)
        client = api_client.get_async_httpx_client()

        response = await client.put(
            f"{config.api_url}/volumes/{volume_id}/files/upload",
            params={"path": path},
            content=content,
            headers={"Content-Type": "application/octet-stream"},
            timeout=config.request_timeout,
        )

        if response.status_code == 404:
            raise NotFoundException(f"Volume '{volume_id}' not found")

        err = handle_api_exception(response)
        if err:
            raise err

        return response.json().get("size", len(content))

    @staticmethod
    async def _download_file(
        volume_id: str,
        path: str,
        **opts: Unpack[ApiParams],
    ) -> bytes:
        """
        Download file content from volume.

        :param volume_id: Volume ID (vol_xxx)
        :param path: File path in volume
        :return: File content as bytes
        """
        config = ConnectionConfig(**opts)
        api_client = get_api_client(config)
        client = api_client.get_async_httpx_client()

        response = await client.get(
            f"{config.api_url}/volumes/{volume_id}/files/download",
            params={"path": path},
            timeout=config.request_timeout,
        )

        if response.status_code == 404:
            raise NotFoundException(
                f"Volume '{volume_id}' or file '{path}' not found"
            )

        err = handle_api_exception(response)
        if err:
            raise err

        return response.content

    @staticmethod
    async def _delete_file(
        volume_id: str,
        path: str,
        recursive: bool = False,
        **opts: Unpack[ApiParams],
    ) -> bool:
        """
        Delete file or directory from volume.

        :param volume_id: Volume ID (vol_xxx)
        :param path: Path to delete
        :param recursive: Delete directory recursively
        :return: True if deleted
        """
        config = ConnectionConfig(**opts)
        api_client = get_api_client(config)
        client = api_client.get_async_httpx_client()

        params: dict = {"path": path}
        if recursive:
            params["recursive"] = "true"

        response = await client.delete(
            f"{config.api_url}/volumes/{volume_id}/files",
            params=params,
            timeout=config.request_timeout,
        )

        if response.status_code == 404:
            raise NotFoundException(
                f"Volume '{volume_id}' or path '{path}' not found"
            )

        err = handle_api_exception(response)
        if err:
            raise err

        return True

File: packages/js-sdk/src/volume/volumeApi.ts
import { ConnectionConfig, ConnectionOpts } from '../connectionConfig'
import { NotFoundError } from '../errors'
import { FileInfo, VolumeInfo } from './types'

/**
 * Allowed mount path prefixes for sandbox volume attachment.
 */
export const ALLOWED_MOUNT_PREFIXES = [
  '/workspace/',
  '/data/',
  '/mnt/',
  '/volumes/',
]

/**
 * Validate volume mount path.
 *
 * @param path Mount path to validate.
 * @throws Error if path is invalid.
 */
export function validateMountPath(path: string): void {
  if (!path) {
    throw new Error('Mount path cannot be empty')
  }

  if (!path.startsWith('/')) {
    throw new Error('Mount path must be absolute (start with /)')
  }

  // Check allowed prefixes
  if (!ALLOWED_MOUNT_PREFIXES.some((prefix) => path.startsWith(prefix))) {
    throw new Error(
      `Mount path must start with one of: ${ALLOWED_MOUNT_PREFIXES.join(', ')}`
    )
  }

  // Check for directory traversal
  if (path.includes('..')) {
    throw new Error("Mount path cannot contain '..'")
  }
}

/**
 * Options for volume API requests.
 */
export interface VolumeApiOpts
  extends Partial<
    Pick<
      ConnectionOpts,
      'apiKey' | 'headers' | 'debug' | 'domain' | 'requestTimeoutMs'
    >
  > {}

interface ApiVolumeInfo {
  volumeID: string
  name: string
  totalSizeBytes?: number
  totalFileCount?: number
  createdAt: string
  updatedAt: string
}

interface ApiFileInfo {
  name: string
  path: string
  type: 'file' | 'directory'
  size?: number
  modifiedAt?: string
}

/**
 * Base class for volume API operations.
 *
 * Uses raw fetch since volume endpoints are not yet in the generated OpenAPI client.
 */
export class VolumeApi {
  protected constructor() {}

  private static async request<T>(
    config: ConnectionConfig,
    method: string,
    path: string,
    opts?: {
      body?: unknown
      signal?: AbortSignal
    }
  ): Promise<T> {
    const url = `${config.apiUrl}${path}`
    const response = await fetch(url, {
      method,
      headers: {
        ...config.headers,
        'Content-Type': 'application/json',
        ...(config.apiKey && { 'X-API-KEY': config.apiKey }),
      },
      body: opts?.body ? JSON.stringify(opts.body) : undefined,
      signal: opts?.signal,
    })

    if (response.status === 404) {
      throw new NotFoundError(`Resource not found: ${path}`)
    }

    if (!response.ok) {
      const text = await response.text()
      throw new Error(`API request failed: ${response.status} ${response.statusText} - ${text}`)
    }

    // Handle empty responses
    const text = await response.text()
    if (!text) {
      return {} as T
    }

    return JSON.parse(text) as T
  }

  /**
   * Create a new volume (idempotent).
   *
   * @param name Volume name.
   * @param opts Connection options.
   * @returns Volume info.
   */
  static async createVolume(
    name: string,
    opts?: VolumeApiOpts
  ): Promise<VolumeInfo> {
    const config = new ConnectionConfig(opts)

    const data = await VolumeApi.request<ApiVolumeInfo>(
      config,
      'POST',
      '/volumes',
      {
        body: { name },
        signal: config.getSignal(opts?.requestTimeoutMs),
      }
    )

    return VolumeApi.parseVolumeInfo(data)
  }

  /**
   * Get volume by ID or name.
   *
   * @param volumeIdOrName Volume ID (vol_xxx) or name.
   * @param opts Connection options.
   * @returns Volume info.
   */
  static async getVolume(
    volumeIdOrName: string,
    opts?: VolumeApiOpts
  ): Promise<VolumeInfo> {
    const config = new ConnectionConfig(opts)

    try {
      const data = await VolumeApi.request<ApiVolumeInfo>(
        config,
        'GET',
        `/volumes/${encodeURIComponent(volumeIdOrName)}`,
        {
          signal: config.getSignal(opts?.requestTimeoutMs),
        }
      )

      return VolumeApi.parseVolumeInfo(data)
    } catch (err) {
      if (err instanceof NotFoundError) {
        throw new NotFoundError(`Volume '${volumeIdOrName}' not found`)
      }
      throw err
    }
  }

  /**
   * List all volumes.
   *
   * @param opts Connection options.
   * @returns List of volume info objects.
   */
  static async listVolumes(opts?: VolumeApiOpts): Promise<VolumeInfo[]> {
    const config = new ConnectionConfig(opts)

    const data = await VolumeApi.request<ApiVolumeInfo[]>(
      config,
      'GET',
      '/volumes',
      {
        signal: config.getSignal(opts?.requestTimeoutMs),
      }
    )

    return (data ?? []).map(VolumeApi.parseVolumeInfo)
  }

  /**
   * Delete a volume.
   *
   * @param volumeIdOrName Volume ID (vol_xxx) or name.
   * @param opts Connection options.
   * @returns True if deleted, false if not found.
   */
  static async deleteVolume(
    volumeIdOrName: string,
    opts?: VolumeApiOpts
  ): Promise<boolean> {
    const config = new ConnectionConfig(opts)

    try {
      await VolumeApi.request<void>(
        config,
        'DELETE',
        `/volumes/${encodeURIComponent(volumeIdOrName)}`,
        {
          signal: config.getSignal(opts?.requestTimeoutMs),
        }
      )
      return true
    } catch (err) {
      if (err instanceof NotFoundError) {
        return false
      }
      throw err
    }
  }

  /**
   * List files in a volume.
   *
   * @param volumeId Volume ID (vol_xxx).
   * @param path Directory path to list.
   * @param opts Connection options.
   * @returns List of file info objects.
   */
  static async listFiles(
    volumeId: string,
    path: string = '/',
    opts?: VolumeApiOpts
  ): Promise<FileInfo[]> {
    const config = new ConnectionConfig(opts)

    try {
      const data = await VolumeApi.request<{ files: ApiFileInfo[] }>(
        config,
        'GET',
        `/volumes/${encodeURIComponent(volumeId)}/files?path=${encodeURIComponent(path)}`,
        {
          signal: config.getSignal(opts?.requestTimeoutMs),
        }
      )

      return (data?.files ?? []).map(VolumeApi.parseFileInfo)
    } catch (err) {
      if (err instanceof NotFoundError) {
        throw new NotFoundError(`Volume '${volumeId}' not found`)
      }
      throw err
    }
  }

  /**
   * Upload file content to volume.
   *
   * @param volumeId Volume ID (vol_xxx).
   * @param path Destination path in volume.
   * @param content File content as Buffer, ArrayBuffer, or ReadableStream.
   * @param opts Connection options.
   * @returns Size of uploaded file.
   */
  static async uploadFile(
    volumeId: string,
    path: string,
    content: Buffer | ArrayBuffer | ReadableStream<Uint8Array>,
    opts?: VolumeApiOpts
  ): Promise<number> {
    const config = new ConnectionConfig(opts)
    const url = `${config.apiUrl}/volumes/${encodeURIComponent(volumeId)}/files/upload?path=${encodeURIComponent(path)}`

    // Prepare headers - for streams, we use chunked transfer encoding
    const headers: Record<string, string> = {
      ...config.headers,
      'Content-Type': 'application/octet-stream',
    }

    // ReadableStream uses chunked transfer encoding (no Content-Length needed)
    // Buffer and ArrayBuffer can include Content-Length for efficiency
    if (content instanceof Buffer) {
      headers['Content-Length'] = String(content.byteLength)
    } else if (content instanceof ArrayBuffer) {
      headers['Content-Length'] = String(content.byteLength)
    }

    const response = await fetch(url, {
      method: 'PUT',
      headers,
      body: content,
      signal: config.getSignal(opts?.requestTimeoutMs),
      // Enable streaming for ReadableStream - required for Node.js fetch
      ...(content instanceof ReadableStream && { duplex: 'half' }),
    } as RequestInit)

    if (response.status === 404) {
      throw new NotFoundError(`Volume '${volumeId}' not found`)
    }

    if (!response.ok) {
      const text = await response.text()
      throw new Error(`Upload failed: ${response.status} ${response.statusText} - ${text}`)
    }

    const data = await response.json() as { size?: number }
    // For ReadableStream, we can only know the size from the server response
    if (data?.size !== undefined) {
      return data.size
    }
    if (content instanceof Buffer) {
      return content.byteLength
    }
    if (content instanceof ArrayBuffer) {
      return content.byteLength
    }
    // For ReadableStream without server response, return 0
    return 0
  }

  /**
   * Download file content from volume.
   *
   * @param volumeId Volume ID (vol_xxx).
   * @param path File path in volume.
   * @param opts Connection options.
   * @returns File content as Buffer.
   */
  static async downloadFile(
    volumeId: string,
    path: string,
    opts?: VolumeApiOpts
  ): Promise<Buffer> {
    const config = new ConnectionConfig(opts)
    const url = `${config.apiUrl}/volumes/${encodeURIComponent(volumeId)}/files/download?path=${encodeURIComponent(path)}`

    const response = await fetch(url, {
      method: 'GET',
      headers: config.headers,
      signal: config.getSignal(opts?.requestTimeoutMs),
    })

    if (response.status === 404) {
      throw new NotFoundError(
        `Volume '${volumeId}' or file '${path}' not found`
      )
    }

    if (!response.ok) {
      throw new Error(`Download failed: ${response.status} ${response.statusText}`)
    }

    const arrayBuffer = await response.arrayBuffer()
    return Buffer.from(arrayBuffer)
  }

  /**
   * Delete file or directory from volume.
   *
   * @param volumeId Volume ID (vol_xxx).
   * @param path Path to delete.
   * @param recursive Delete directory recursively.
   * @param opts Connection options.
   * @returns True if deleted.
   */
  static async deleteFile(
    volumeId: string,
    path: string,
    recursive: boolean = false,
    opts?: VolumeApiOpts
  ): Promise<boolean> {
    const config = new ConnectionConfig(opts)
    const url = `/volumes/${encodeURIComponent(volumeId)}/files?path=${encodeURIComponent(path)}&recursive=${recursive}`

    try {
      await VolumeApi.request<void>(
        config,
        'DELETE',
        url,
        {
          signal: config.getSignal(opts?.requestTimeoutMs),
        }
      )
      return true
    } catch (err) {
      if (err instanceof NotFoundError) {
        throw new NotFoundError(
          `Volume '${volumeId}' or path '${path}' not found`
        )
      }
      throw err
    }
  }

  private static parseVolumeInfo(data: ApiVolumeInfo): VolumeInfo {
    return {
      volumeId: data.volumeID,
      name: data.name,
      totalSizeBytes: data.totalSizeBytes ?? 0,
      totalFileCount: data.totalFileCount ?? 0,
      createdAt: new Date(data.createdAt),
      updatedAt: new Date(data.updatedAt),
    }
  }

  private static parseFileInfo(data: ApiFileInfo): FileInfo {
    return {
      name: data.name,
      path: data.path,
      type: data.type,
      size: data.size,
      modifiedAt: data.modifiedAt ? new Date(data.modifiedAt) : undefined,
    }
  }
}
